# 🧠 NLP Workshop: IR Inverted Matrix

## 👥 Group Information

**Group Name:** Group 8  

**Members:**
- Haibo Yuan (Student ID: 9010929)  
- Abdalla Mohamed (Student ID: 9089339)

---

# 🛠️ Active Learning Workshop: Implementing an Inverted Matrix (Jupyter + GitHub Edition)
## 🔍 Workshop Theme
*Readable, correct, and collaboratively reviewed code—just like in the real world.*


Welcome to the 90-minute workshop! In this hands-on session, your team will build an **Inverted Index** pipeline, the foundation of many intelligent systems that need fast and relevant access to text data — such as AI agents.

### 👥 Team Guidelines
- Work in teams of 3.
- Submit one completed Jupyter Notebook per team.
- The final notebook must contain **Markdown explanations** and **Python code**.
- Push your notebook to GitHub and share the `.git` link before class ends.

---
## 🔧 Workshop Tasks Overview

1. **Document Collection**
2. **Tokenizer Implementation**
3. **Normalization Pipeline (Stemming, Stop Words, etc.)**
4. **Build and Query the Inverted Index**

> Each step includes a sample **talking point**. Your team must add your own custom **Markdown + code cells** with a **second talking point**, and test your Inverted Index with **2 phrase queries**.


## 📝 Team Introduction

In this workshop, our team (Group 8) aims to implement an inverted index pipeline step by step, including document collection, tokenization, normalization, and query processing.

The goal of this workshop is to understand how search engines and AI systems efficiently retrieve relevant information from large text datasets. By building this system ourselves, we gain practical experience in Natural Language Processing (NLP) and Information Retrieval (IR).

We will also test our implementation using different queries to evaluate how well the inverted index performs.

---



## 🧠 Learning Objectives
- Implement an **Inverted Matrix** using real-world data during the NLP process.
- Build **Jupyter Notebooks** with well-structured code and clear Markdown documentation.
- Use **Git and GitHub** for collaborative version control and code sharing.
- Identify and articulate coding issues ("**talking points**") and insert them directly into peer notebooks.
- Practice **collaborative debugging**, professional peer feedback, and improve code quality.

## 🧩 Workshop Structure (90 Minutes)
1. **Instructor Use Case Introduction** *(15 min)* – Set up teams of 3 people. Read and understand the workshop, plus submission instructions. Seek assistance if needed.
2. **Team Jupyter Notebook Development** *(45 min)* – Manual IR and Inverted Matrix coding + Markdown documentation (work as teams)
3. **Push to GitHub** *(15 min)* – Teams commit and push initial notebooks. **Make sure to include your names so it is easy to identify the team that developed the Min-Max code**.
4. **Instructor Review** - The instructor will go around, take notes, and provide coaching as needed, during the **Peer Review Round**
5. **Email Delivery** *(15 min)* – Each team send the instructor an email **with the *.git link** to the GitHub repo **(one email/team)**. Subject on the email is: PROG8245 - Inverted Matrix  Workshop, Team #_____.
6. **Push to the course GitHub** *(automated)* – Find the last code cell in this notebook, update your team number, and run the cell. It will push your notebook to the repo at `submissions/team#`

## 💻 Submission Checklist
- ✅ `IR_InvertedMatrix_Workshop.ipynb` with:
  - Demo code: Document Collection, Tokenizer, Normalization Pipeline, and Inverted Index.
  - Markdown explanations for each major step
  - **Labeled talking point(s)** and 2 phrase query tests
- ✅ `README.md` with:
  - Dataset description
  - Team member names
  - Link to the dataset and license (if public)
- ✅ GitHub Repo:
  - Public repo named `IR-invertedmatrix-workshop`
  - This is a group effort, so **choose one member of the team** to publish the repo
  - At least **one commit containing one meaningful talking point**

## 📝 Team Understanding of Workshop Requirements

Based on the workshop instructions, our team understands that we need to:

- Implement the full Inverted Index pipeline, including document collection, tokenization, normalization, and querying.
- Provide clear Markdown explanations for each step to demonstrate our understanding.
- Include at least one meaningful **talking point** that reflects challenges, observations, or improvements.
- Test the system using at least **two phrase queries** to validate the correctness of our inverted index.
- Ensure the notebook runs completely without errors before submission.

We will follow these requirements carefully to ensure both correctness and completeness of our work.

---

## 📄 Step 1: Document Collection


### 🗣 Instructor Talking Point:
> We begin by gathering a text corpus. To build a robust index, your vocabulary should include **over 2000 unique words**. You can use scraped articles, academic papers, or open datasets.

### 🔧 Your Task:
- Collect at least 20+ text documents.
- Ensure the vocabulary exceeds 2000 unique words.
- Load the documents into a list for processing.


## 📝 Team Talking Point – Document Collection

In this step, we collected a set of text documents to build our inverted index. Having a sufficiently large and diverse dataset is important because it ensures that the vocabulary is rich and representative of real-world language usage.

We used the provided sample documents, which were automatically downloaded from an RSS feed. This approach simulates real-world data collection, where documents are gathered from external sources.

A key challenge in this step is ensuring that the dataset is large enough (at least 20 documents) and contains a wide range of words. A richer vocabulary leads to a more effective and meaningful inverted index.

---

In [226]:
import os
import requests
import feedparser

# Example: Download blog posts from a public RSS feed and save as .txt files
def fetch_and_save_blog_posts(rss_url, save_folder, max_posts=20):
    feed = feedparser.parse(rss_url)
    if not os.path.exists(save_folder):
        os.makedirs(save_folder)
    count = 0
    for entry in feed.entries:
        if count >= max_posts:
            break
        title = entry.get('title', 'untitled')
        content = entry.get('summary', '')
        # Clean filename
        filename = f"blog_{count+1}.txt"
        filepath = os.path.join(save_folder, filename)
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(f"{title}\n\n{content}")
        count += 1
    print(f"Saved {count} blog posts to {save_folder}")

# Example RSS feed: Python Software Foundation blog
rss_url = 'https://pyfound.blogspot.com/feeds/posts/default?alt=rss'
fetch_and_save_blog_posts(rss_url, 'sample_docs', max_posts=20)

Saved 20 blog posts to sample_docs


In [227]:
# Example: Load text files from a folder
import os

def load_documents(folder_path):
    documents = []
    for filename in os.listdir(folder_path):
        if filename.endswith('.txt'):
            with open(os.path.join(folder_path, filename), 'r', encoding='utf-8') as file:
                documents.append(file.read())
    return documents

# Replace 'sample_docs/' with your actual folder
documents = load_documents('sample_docs/')
print(f"Loaded {len(documents)} documents.")


Loaded 20 documents.


## 📝 Team Talking Point – Document Loading

In this step, we loaded all text documents from the local folder into a Python list. This allows us to process the data efficiently in later stages such as tokenization and normalization.

Storing documents in a list makes it easier to iterate through them and apply transformations step by step in the NLP pipeline.

---

## ✂️ Step 2: Tokenizer


### 🗣 Instructor Talking Point:
> The tokenizer breaks raw text into a stream of words (tokens). This is the foundation for every later step in IR and NLP.

### 🔧 Your Task:
- Implement a basic tokenizer that splits text into lowercase words.
- Handle punctuation removal and basic non-alphanumeric filtering.


In [228]:
import re

def tokenize(text):
    tokens = re.findall(r'\b\w+\b', text.lower())
    return tokens

# Test on one document
tokens = tokenize(documents[0])
print(tokens[:20])  # Preview first 20 tokens


['applications', 'to', 'join', 'the', 'psf', 'meetup', 'pro', 'network', 'are', 'back', 'open', 'p', 'following', 'the', 'a', 'href', 'https', 'pyfound', 'blogspot', 'com']


## 📝 Team Talking Point – Tokenizer

In this step, we used a basic tokenizer to split raw text into lowercase word tokens. The tokenizer applies a regular expression to extract word-like patterns and removes most punctuation automatically.

This step is important because tokenization converts unstructured text into a format that can be processed in later stages such as normalization and inverted indexing.

A practical issue we observed is that simple tokenization may still keep noisy terms from URLs or HTML-like fragments. However, this lightweight approach is still effective as a first step in the NLP pipeline.

The tokenizer converts the first document into a list of normalized lowercase tokens for preview and inspection.

## 🔁 Step 3: Normalization Pipeline (Stemming, Stop Word Removal, etc.)


### 🗣 Instructor Talking Point:
> Now we normalize tokens: convert to lowercase, remove stop words, apply stemming or affix stripping. This reduces redundancy and enhances search accuracy.

### 🔧 Your Task:
- Use `nltk` to remove stopwords and apply stemming.


In [229]:
import nltk
from nltk.corpus import stopwords
from nltk.stem import PorterStemmer

nltk.download('stopwords')
stop_words = set(stopwords.words('english'))
stemmer = PorterStemmer()

def normalize_tokens(tokens):
    return [stemmer.stem(t) for t in tokens if t not in stop_words]

# Example: normalize one document
norm_tokens = normalize_tokens(tokens)
print(norm_tokens[:20])


['applic', 'join', 'psf', 'meetup', 'pro', 'network', 'back', 'open', 'p', 'follow', 'href', 'http', 'pyfound', 'blogspot', 'com', '2026', '02', 'introduc', 'psf', 'commun']


[nltk_data] Downloading package stopwords to
[nltk_data]     C:\Users\10122\AppData\Roaming\nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


## 📝 Team Talking Point – Normalization Pipeline

In this step, we applied normalization techniques including stop word removal and stemming to clean and standardize the tokens.

Stop words such as "the", "is", and "and" were removed because they do not carry significant meaning for search or indexing. This helps reduce noise in the dataset.

We also applied stemming using the PorterStemmer to reduce words to their root form. For example, words like "running", "runs", and "ran" can be reduced to a common base form. This improves matching during search queries.

One challenge we observed is that stemming can sometimes produce non-intuitive words (e.g., "applications" becomes "applic"), but it is still useful for improving recall in search systems.

---

## 🔍 Step 4: Inverted Index


### 🗣 Instructor Talking Point:
> We now map each normalized token to the list of document IDs in which it appears. This is the core structure that allows fast Boolean and phrase queries.

### 🔧 Your Task:
- Build the inverted index using a dictionary.
- Add code to support phrase queries using positional indexing.


In [230]:
from collections import defaultdict

def build_inverted_index(documents):
    index = defaultdict(list)
    for doc_id, text in enumerate(documents):
        tokens = normalize_tokens(tokenize(text))
        seen = set()
        for token in tokens:
            if token not in seen:
                index[token].append(doc_id)
                seen.add(token)
    return dict(index)

inverted_index = build_inverted_index(documents)
print(dict(list(inverted_index.items())[:10]))  # Preview first 10 terms


{'applic': [0, 2, 3, 4, 5, 6, 8, 14, 18], 'join': [0, 2, 3, 4, 7, 10, 11, 13, 14, 15, 16, 18], 'psf': [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19], 'meetup': [0, 3, 13, 14], 'pro': [0, 2], 'network': [0, 3, 15], 'back': [0, 3, 4, 8, 10, 13, 14], 'open': [0, 3, 4, 6, 8, 10, 11, 12, 13, 18, 19], 'p': [0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19], 'follow': [0, 6, 7, 9, 11, 13, 14, 17, 19]}


## 📝 Team Talking Point – Inverted Index

In this step, we built an inverted index that maps each token to a list of document IDs where the token appears. This structure is essential for efficient information retrieval.

By using a dictionary, we can quickly look up which documents contain a given word. This significantly improves search performance compared to scanning all documents.

We also ensured that each document ID appears only once per token by using a set to track seen tokens. This avoids redundancy in the index.

One limitation of this basic inverted index is that it does not store word positions, which makes phrase queries difficult. This motivates the use of positional indexing for more advanced search capabilities.

---

## 🧪 Test: Phrase Queries


### 🗣 Instructor Talking Point:
> A phrase query requires the exact sequence of terms (e.g., "machine learning"). To support this, extend the inverted index to store positions, not just docIDs.

### 🔧 Your Task:
- Implement 2 phrase queries.
- Demonstrate that they return the correct documents.


In [231]:
# Placeholder for phrase query implementation
# You may build a position-aware index or use string search within docs after normalization

# Example:
query1 = "Python software"
query2 = "bylaws"

def phrase_query(query, documents):
    results = []
    query_tokens = normalize_tokens(tokenize(query))
    for doc_id, doc in enumerate(documents):
        doc_tokens = normalize_tokens(tokenize(doc))
        # Search for the sequence in the token list
        for i in range(len(doc_tokens) - len(query_tokens) + 1):
            if doc_tokens[i:i+len(query_tokens)] == query_tokens:
                results.append(doc_id)
                break
    return results

# Run phrase queries on the loaded documents
results1 = phrase_query(query1, documents)
results2 = phrase_query(query2, documents)

print(f"Documents containing the phrase '{query1}': {results1}")
print(f"Documents containing the phrase '{query2}': {results2}")

Documents containing the phrase 'Python software': [0, 3, 4, 5, 6, 8, 9, 10, 11, 13, 14, 15, 16, 17, 18, 19]
Documents containing the phrase 'bylaws': []


## 📝 Team Talking Point – Phrase Queries

In this step, we implemented phrase queries to search for exact sequences of words within documents. Unlike single-term queries, phrase queries require that words appear in the correct order and consecutively.

We used a sliding window approach to compare sequences of tokens in each document with the query tokens. This allows us to identify documents that contain the exact phrase.

One limitation of this approach is that it scans each document sequentially, which can be inefficient for large datasets. A more advanced solution would involve positional indexing to improve performance.

---

## 🧠 Additional Challenge: Implement Positional Indexes

Implement Positional Indexes, an advanced version of an inverted index that not only stores which documents a term appears in, but also where (at what positions) the term occurs within each document.

Apply it to the documents you used to produce the Inverted Matrix during the Active Learning Workshop. 

Then compare it against a **Term Document Count Matrix** (which you don't have to implement)

And finally, fill in the blanks in the table below.

| Term | What is it? | How is it used? | Pros | Cons |
|----------------|-------------|-------------|-------------|-------------|
| Term Frequency (TF) | TF measures how many times a term appears in a document. | It is used to estimate how important a word is within a specific document. | Simple to compute; highlights frequently used words in a document. | Common words may get high scores even if they are not informative. |
| Inverse Document Frequency (idF weight) | IDF measures how rare a term is across all documents. | It is used to reduce the importance of common words and emphasize rare terms. | Improves search relevance by highlighting distinctive words. | Requires global corpus statistics; rare words are not always meaningful. |

Then use the table to prepare talking points on:
- Which implementation you prefer to use for searching bigrams (a.k.a., biwords), pairs of consecutive words in a document, and why?

#### Sample solution:

In [232]:
# Implement Positional Indexes: Store term positions for each document
from collections import defaultdict

def build_positional_index(documents):
    positional_index = defaultdict(lambda: defaultdict(list))
    for doc_id, text in enumerate(documents):
        tokens = normalize_tokens(tokenize(text))
        for pos, token in enumerate(tokens):
            positional_index[token][doc_id].append(pos)
    return positional_index

# Build positional index for the loaded documents
positional_index = build_positional_index(documents)

# Example: Show positions for a sample term
sample_term = 'python'  # Change to any term you want to inspect
print(f"Positions for term '{sample_term}':")
for doc_id, positions in positional_index[sample_term].items():
    print(f"  Document {doc_id}: {positions}")

Positions for term 'python':
  Document 0: [29, 37, 47, 61, 93, 163, 233, 264, 275, 350, 395, 429, 507, 515, 521, 532, 541, 545, 558, 564, 575, 584]
  Document 1: [13, 29, 35, 37, 48, 61, 78, 80, 137, 139, 164]
  Document 2: [0, 18, 33, 48, 52, 57, 105, 118, 145, 225, 258, 292, 321, 326, 346, 416, 470, 478, 508, 516, 538, 543, 554, 566, 584, 591, 623, 636, 642]
  Document 3: [0, 12, 17, 24, 29, 47, 73, 81, 93, 104, 121, 128, 147, 150, 171, 248, 294, 302, 331, 339, 359, 370, 382, 400, 406, 435, 450, 457, 537, 567, 613, 626, 652, 758, 813, 864, 901, 1024, 1060, 1067, 1082, 1087, 1099, 1111, 1135]
  Document 4: [83, 293, 376, 529, 840, 879, 891, 911, 917, 953, 960, 982, 1013, 1016, 1066, 1140, 1156]
  Document 5: [3, 9, 18, 28, 72, 97, 133, 153, 156, 269, 292, 372, 382, 392, 399, 409, 441, 448, 451, 458]
  Document 6: [4, 10, 60, 88, 263, 435, 501, 634, 870, 1051, 1061, 1071, 1078, 1088, 1109]
  Document 7: [18, 28, 80, 149, 156, 183, 210, 426, 441, 463]
  Document 8: [48, 227, 245, 251, 

## 📝 Team Talking Point – Bigram Search

For searching bigrams (pairs of consecutive words), we prefer using a positional index rather than a simple inverted index or term-document count matrix.

A positional index stores the exact positions of terms in each document, which allows us to verify whether two words appear next to each other in the correct order. This makes phrase and bigram queries more accurate.

Although positional indexes require more storage and processing time, they provide significantly better precision for phrase-based search tasks compared to basic indexing methods.

### 📊 Comparison: Positional Index vs. Term Document Count Matrix

A **Positional Index** stores not only which documents a term appears in, but also the exact positions of each term within those documents. In contrast, a **Term Document Count Matrix** (TDCM) simply records the frequency of each term in each document, without any positional information.

| Feature                      | Positional Index                | Term Document Count Matrix (TDCM) |
|------------------------------|---------------------------------|-----------------------------------|
| Stores term positions        | Yes                             | No                                |
| Supports phrase/bigram search| Yes                             | No                                |
| Memory usage                 | Higher                          | Lower                             |
| Query speed for phrases      | Fast                            | Slow (requires scanning)          |
| Useful for                   | Phrase queries, proximity search | Keyword frequency analysis         |
| Implementation complexity    | Higher                          | Lower                             |

**Summary:**
- Use a positional index for advanced search features like phrase and proximity queries.
- Use a TDCM for simple keyword frequency analysis and ranking.

### 📝 Completed Table: Term Frequency (TF) and Inverse Document Frequency (IDF weight)

| Term                           | What is it?                                                                 | How is it used?                          | Pros                                   | Cons                                   |
|-------------------------------|----------------------------------------------------------------------------|------------------------------------------|----------------------------------------|----------------------------------------|
| Term Frequency (TF)            | Number of times a term appears in a document                               | Measures term importance in a document   | Simple, fast, good for ranking         | Doesn't account for common terms       |
| Inverse Document Frequency     | Log ratio of total docs to docs containing the term                        | Weights rare terms higher                | Reduces impact of common words         | Can overweight rare, irrelevant terms  |
| (IDF weight)                   | Used with TF to score term relevance: TF-IDF = TF * IDF                    | Improves search relevance                | Highlights unique, important terms     | Requires global corpus statistics      |

Let's implement the Inverse Document Frequency (IDF)

In [233]:
# Calculate Inverse Document Frequency (IDF) for each term using the positional index
import math

def compute_idf_from_positional_index(positional_index, total_docs):
    idf_scores = {}
    for term, doc_positions in positional_index.items():
        doc_count = len(doc_positions)
        if doc_count > 0:
            # --- IGNORE ---
            # print(f"Computing IDF for {term} '{total_docs}' / {doc_count}")
            idf_scores[term] = math.log(total_docs / doc_count)
        else:
            idf_scores[term] = 0.0
    return idf_scores

# Compute IDF for all terms using the positional index
idf_scores = compute_idf_from_positional_index(positional_index, len(documents))

# Example: Show IDF for a sample term
# NOTE : We applied stemming in the normalization step, so the term may not match exactly if it was stemmed
sample_term = 'softwar'  # Change to any term you want to inspect
print(f"IDF for term '{sample_term}': {idf_scores.get(sample_term, 0.0)}")

IDF for term 'softwar': 0.22314355131420976


## 📝 Team Talking Point – IDF (Inverse Document Frequency)

In this step, we implemented the IDF calculation to measure how rare a term is across the document collection.

IDF assigns higher values to terms that appear in fewer documents, making them more important for distinguishing relevant content. Common words that appear in many documents receive lower IDF scores.

We used the formula:

IDF = log(total_documents / document_frequency)

This improves search relevance when combined with TF in TF-IDF. However, IDF may overemphasize rare but unimportant terms and requires knowledge of the entire dataset.

### 📐 Mathematical Explanation: Inverse Document Frequency (IDF)

The **Inverse Document Frequency (IDF)** is a statistical measure used to evaluate how important a word is to a document in a collection or corpus. The intuition is that terms that appear in many documents are less informative than those that appear in few.

The IDF for a term $t$ is defined as:

$$
\text{IDF}(t) = \log\left(\frac{N}{n_t}\right)
$$

Where:
- $N$ is the total number of documents in the corpus.
- $n_t$ is the number of documents containing the term $t$.

---

#### Example Calculation

Suppose:
- $N = 20$ (total documents)
- $n_{\text{softwar}} = 15$ (documents containing the term 'softwar')

Then:

$$
\text{IDF}(\text{softwar}) = \log\left(\frac{20}{15}\right) \approx 0.2877
$$

So, the IDF for term 'softwar' is $0.2877$.

A higher IDF value means the term is rare across the corpus, while a lower value means the term is common. IDF is often used in combination with Term Frequency (TF) to compute the TF-IDF score:

$$
\text{TF-IDF}(t, d) = \text{TF}(t, d) \times \text{IDF}(t)
$$

Where $\text{TF}(t, d)$ is the frequency of term $t$ in document $d$.

This weighting helps highlight terms that are both frequent in a document and rare across the corpus, improving search relevance and document ranking.

In [234]:
# Calculate Term Frequency (TF) for the term 'softwar' in each document
# TF is simply the number of times the term appears in a document
import pandas as pd

term = 'softwar'  # Use the stemmed term as before

tf_values = []
for doc_id, doc in enumerate(documents):
    tokens = normalize_tokens(tokenize(doc))
    tf = tokens.count(term)
    tf_values.append(tf)
    print(f"TF for term '{term}' in Document {doc_id}: {tf}")

# Display as a DataFrame for clarity
TF_df = pd.DataFrame({'Document': list(range(len(documents))), 'TF': tf_values})
print(TF_df)

TF for term 'softwar' in Document 0: 6
TF for term 'softwar' in Document 1: 0
TF for term 'softwar' in Document 2: 0
TF for term 'softwar' in Document 3: 3
TF for term 'softwar' in Document 4: 1
TF for term 'softwar' in Document 5: 2
TF for term 'softwar' in Document 6: 4
TF for term 'softwar' in Document 7: 0
TF for term 'softwar' in Document 8: 2
TF for term 'softwar' in Document 9: 1
TF for term 'softwar' in Document 10: 3
TF for term 'softwar' in Document 11: 3
TF for term 'softwar' in Document 12: 0
TF for term 'softwar' in Document 13: 4
TF for term 'softwar' in Document 14: 3
TF for term 'softwar' in Document 15: 1
TF for term 'softwar' in Document 16: 2
TF for term 'softwar' in Document 17: 1
TF for term 'softwar' in Document 18: 3
TF for term 'softwar' in Document 19: 6
    Document  TF
0          0   6
1          1   0
2          2   0
3          3   3
4          4   1
5          5   2
6          6   4
7          7   0
8          8   2
9          9   1
10        10   3
11    

#### 📊 Example: Term Frequency (TF) and TF-IDF Calculation for 'softwar' in Document 0

TF for term 'softwar' in Document 0:

$$
\text{TF}(\text{softwar}, 0) = n_{\text{softwar}, 0}
$$

TF-IDF for term 'softwar' in Document 0:

$$
\text{TF-IDF}(\text{softwar}, 0) = \text{TF}(\text{softwar}, 0) \times \text{IDF}(\text{softwar}) = n_{\text{softwar}, 0} \times \text{IDF}(\text{softwar})
$$

Where:
- $n_{\text{softwar}, 0}$ is the number of times 'softwar' appears in Document 0.
- $\text{IDF}(\text{softwar})$ is the previously calculated IDF value for 'softwar'.

Substitute the actual values from your output to see the final TF-IDF score for Document 0.

## 📝 Team Talking Point – TF-IDF Example

From the output above, we can observe that the term "softwar" appears **6 times** in Document 0.

Previously, we computed the IDF value for "softwar" as approximately **0.2231**.

Using the TF-IDF formula:

TF-IDF = TF × IDF

For Document 0:

TF-IDF = 6 × 0.2231 ≈ 1.3386

This shows that although "softwar" appears frequently in Document 0, its overall importance is moderated by its presence across multiple documents.

TF-IDF helps balance term frequency with term rarity, making it useful for identifying important terms for search and ranking tasks.

In [235]:
def query_processor_tfidf(query, documents, idf_scores):
    results = []

    for doc_id, doc in enumerate(documents):
        tokens = normalize_tokens(tokenize(doc))
        score = 0

        for term in query.split():
            tf = tokens.count(term)
            idf = idf_scores.get(term, 0)
            score += tf * idf

        if score > 0:
            results.append((doc_id, score))

    # sort by score (descending)
    results.sort(key=lambda x: x[1], reverse=True)

    return results

## 📝 Reflection

In this section, we extended a basic query processor to incorporate TF-IDF scoring for ranking documents.

The original approach only returned documents containing the query terms, without considering their importance. By using TF-IDF, we can rank documents based on both term frequency and term rarity.

This improves search relevance by prioritizing documents where the query terms are both frequent and distinctive.

One limitation of this approach is that it still treats terms independently and does not consider word order or phrase structure, which can be addressed using positional indexes.

## 🧠 Push your code to the course GitHub repository

Update your team number and then run the code. 

In [237]:
!python submit_assignment.py --notebook IR_InvertedMatrix_Workshop.ipynb --student_id "group8"


Traceback (most recent call last):
  File "d:\Desktop\8245\IRInvertedMatrix_Workshop\submit_assignment.py", line 18, in <module>
    from git import Repo, GitCommandError
ModuleNotFoundError: No module named 'git'
